src/twisted/protocols/test/test_tls.py

```diff
@@ -16,19 +16,20 @@
 try:
     from twisted.protocols.tls import TLSMemoryBIOProtocol, TLSMemoryBIOFactory
     from twisted.protocols.tls import _PullToPush, _ProducerMembrane
-    from OpenSSL.crypto import X509Type
+    from OpenSSL import crypto
     from OpenSSL.SSL import (
         TLSv1_METHOD,
         TLSv1_1_METHOD,
         TLSv1_2_METHOD,
         Error,
         Context,
-        ConnectionType,
+        Connection,
         WantReadError,
     )
+    from twisted.internet._sslverify import _setAcceptableProtocols
 except ImportError:
     # Skip the whole test module if it can't be imported.
-    skip = "pyOpenSSL 0.10 or newer required for twisted.protocol.tls"
+    skip = "pyOpenSSL 16.0.0 or newer required for twisted.protocol.tls"
```

Comment:
I suppose this is the anti-pattern that led to the ticket being resolved in the first place.

This shows up in Twisted over and over so it's hardly a problem specific to either this module or this PR (and this PR is only changing some of the details of the pre-existing instance of this anti-pattern).

So.  This comment is not a blocker.  That said, I think a good replacement would be a pattern where the skip only happens if the dependency isn't met, not if *any* `ImportError` is raised.  It would be nice if this case fit into the model that the various skip decorators provider... but since there are possibly failing imports it's not clear that it does.  Or, I dunno, what about:

```
from OpenSSL import __version__ as pyOpenSSLVersion
try:
    from OpenSSL.SSL import (
        ...
    )
except ImportError:
    pass

@skipIf(pyOpenSSLVersion < "16.0.0", "pyOpenSSL 16.0.0 or newer required for twisted.protocol.tls")
class SomeSSLTests(...):
    ...
```

If the version check is the correct one to make then the tests won't run and it doesn't matter if the imports succeeded or not.

If the version check is out of step with reality then the tests will fail and someone can look at the problem.  That's exactly what we would have wanted to happen when the first `ConnectionType`-lacking pyOpenSSL release came out, I think.

I guess in this case we want to skip the whole module worth of tests so maybe this actually ends up looking like:

```
from OpenSSL import __version__ as pyOpenSSLVersion
try:
    from OpenSSL.SSL import (
        ...
    )
except ImportError:
    pass

if pyOpenSSLVersion < "16.0.0":
    skip ="pyOpenSSL 16.0.0 or newer required for twisted.protocol.tls"
```

Reply:
This looks like a great idea to me. I filed a ticket to make it policy: https://twistedmatrix.com/trac/ticket/10082